File: lib/domain/recovery/recovery_repository.dart
import 'dart:io';

import 'package:buffer/domain/recovery/recovery_note.dart';

/// Domain port (repository interface) for the recovery persistence boundary.
///
/// ## M2 surface (FR-M2-09, §5.1.1)
///
/// The sole M2 member is [save]. The Emergency-Recovery milestone (M5 /
/// Phase 4) extends this SAME interface additively with `list`, `read`,
/// `delete`, `deleteAll`, and `trim` — added below (OCP: new behaviour
/// via additive implementation, not edits to this interface).
///
/// ## Precondition (caller-enforced)
///
/// The caller (the `SaveBufferToRecovery` use case) guarantees that [text]
/// is non-empty after trimming before delegating to [save]. The repository
/// does NOT re-validate emptiness — that guard lives at the use-case boundary.
///
/// ## Error contract
///
/// A [FileSystemException] thrown by the underlying I/O MUST propagate to
/// the caller unchanged. Implementations must NOT swallow, wrap, or convert
/// it; the lifecycle host is responsible for catching and logging it so the
/// app never crashes on backgrounding (EC-M2-08).
///
/// ## Domain purity
///
/// This file imports only `dart:io` for [File] and the co-located
/// [RecoveryNote] domain entity. It must never import the Flutter SDK or any
/// infrastructure package.
abstract interface class RecoveryRepository {
  /// Persists [text] to the recovery store and returns the written [File].
  ///
  /// Precondition: `text.trim().isNotEmpty` — enforced by the use case.
  /// On [FileSystemException]: propagates to the caller unchanged (not
  /// swallowed).
  Future<File> save(String text);

  // --- M5, ADDED (additive; no breaking change) ---

  /// Returns saved notes NEWEST-FIRST. When the recovery directory is absent,
  /// returns const []. Never creates the directory. savedAt parsed from the
  /// filename. Malformed filenames are skipped. Propagates FileSystemException
  /// (UI-path caller catches and renders).
  Future<List<RecoveryNote>> list();

  /// Returns the full UTF-8 text of the note's backing file. Used by restore.
  /// Propagates FileSystemException (UI-path caller catches).
  Future<String> read(RecoveryNote note);

  /// Deletes exactly the one file identified by [note] (siblings untouched).
  /// No-op when the directory or the file is absent. Never creates the
  /// directory. Propagates FileSystemException (UI-path caller catches).
  Future<void> delete(RecoveryNote note);

  /// Deletes every recovery `.txt` file. No-op when the directory is absent.
  /// Never creates the directory. Propagates FileSystemException.
  Future<void> deleteAll();

  /// Retains the newest [keep] files by LEXICOGRAPHIC FILENAME, deletes the
  /// rest. No-op when the directory is absent or file count <= keep. Tie-break
  /// by filename, NEVER mtime. Called as trim(10) from SaveBufferToRecovery
  /// AFTER a successful save. Propagates FileSystemException UNCHANGED like
  /// save (background path — lifecycle host catches/logs).
  Future<void> trim(int keep);

  // --- NEW (additive, Defect B, C-04) ---

  /// Persists [text] to the recovery store SYNCHRONOUSLY and returns the
  /// written [File]. Used on the `paused`/`detached` lifecycle path where the
  /// OS may freeze the isolate before any async I/O flushes. Implementations
  /// MUST complete the directory-create + write + collision-resolution + the
  /// trim-to-keep step before returning (no awaits, no queued microtasks).
  ///
  /// Precondition: `text.trim().isNotEmpty` — enforced by the caller.
  /// On [FileSystemException]: propagates to the caller UNCHANGED (the
  /// lifecycle host catches and logs; never crashes — EC-M2-08).
  /// Retention: trims to the newest [keep] files (default 10) by LEXICOGRAPHIC
  /// filename AFTER the write succeeds, NEVER mtime (NFR-M5-01).
  File saveSync(String text, {int keep = 10});
}

File: shared/src/commonMain/kotlin/com/paolosantucci/foglietto/shared/recovery/RecoveryRepository.kt
package com.paolosantucci.foglietto.shared.recovery

/**
 * Domain port (repository interface) for the recovery persistence boundary.
 *
 * Port of the Dart `RecoveryRepository` abstract interface class.
 *
 * ## API surface (§5.1.1)
 *
 * Six non-suspend members (okio I/O is blocking — the Dart async split collapses
 * to one synchronous save):
 *   - [save]      — writes the buffer text; returns the written file path as a String.
 *   - [list]      — newest-first notes; empty list when the dir is absent.
 *   - [read]      — returns file text or null if the file has vanished.
 *   - [delete]    — no-op if the file or directory is absent.
 *   - [deleteAll] — no-op if the directory is absent.
 *   - [trim]      — keep newest `keep` files by lexicographic filename (NEVER mtime).
 *
 * ## Dropped from the Dart oracle
 *
 *   - `saveSync` — dropped by construction (okio is synchronous; no async split needed).
 *   - `keep` default on `trim` — the literal `10` lives at the use-case boundary
 *     ([SaveBufferToRecovery]), NOT on this interface.
 *
 * ## Error contract
 *
 * okio `IOException` from [save] MUST propagate unchanged (no swallow, no wrap).
 * [list], [read], [delete], [deleteAll], [trim] tolerate an absent directory and
 * vanished/changed files without throwing (FR-16, FR-18, R-A4).
 *
 * ## Domain purity
 *
 * This file imports only types within the `recovery` package. It must never import
 * okio, Flutter, or any infrastructure package.
 *
 * Spec refs: §5.1.1, FR-11..FR-18; assessment R-A3, R-A4, R-A5 (drop-list).
 */
interface RecoveryRepository {

    /**
     * Persists [text] to the recovery store and returns the written file path as a [String].
     *
     * Precondition: `text.trim().isNotEmpty` — enforced by [SaveBufferToRecovery], not here.
     * On okio `IOException`: propagates to the caller unchanged (not swallowed).
     */
    fun save(text: String): String

    /**
     * Returns saved notes NEWEST-FIRST by parsed [RecoveryInstant].
     *
     * When the recovery directory is absent returns an empty list. Never creates the directory.
     * Malformed filenames (not matching the ms grammar) are silently skipped.
     * Tolerates files that have vanished or changed since the directory was listed.
     */
    fun list(): List<RecoveryNote>

    /**
     * Returns the full UTF-8 text of the recovery file at [path].
     *
     * Returns `null` if the file has vanished since listing (Files-app mutability).
     * Never throws on an absent file.
     */
    fun read(path: String): String?

    /**
     * Deletes the recovery file at [path].
     *
     * No-op when the directory or the file is absent. Never creates the directory.
     * Never throws on an absent target.
     */
    fun delete(path: String)

    /**
     * Deletes every recovery `.txt` file.
     *
     * No-op when the directory is absent. Never creates the directory.
     */
    fun deleteAll()

    /**
     * Retains the newest [keep] files ranked by LEXICOGRAPHIC FILENAME, deletes the rest.
     *
     * No-op when the directory is absent or when the file count is at or below [keep].
     * Tie-breaking is by filename, NEVER mtime (FR-14, R-A3).
     *
     * Note: [keep] has NO default value — the literal `10` lives at the use-case
     * boundary ([SaveBufferToRecovery.invoke]), not on this interface.
     */
    fun trim(keep: Int)
}
